modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/webpack_federation.js

```diff
@@ -12,61 +12,198 @@
-				container = Liferay.Webpack.Container[containerId];
+		explain('Fetching container', containerId, 'from', url);
+
+		containerRequests[containerId] = createContainerRequest(
+			url,
+			(containerRequest) => {
+				explain('Fetched container', containerId);
+
+				const container = getContainer(containerId);
 
 				if (container) {
-					Promise.resolve(container.init(Liferay.Webpack.SharedScope))
-						.then(() => container.get(path))
-						.then((module) => resolve(module()));
+					explain(
+						'Initializing container', containerId, '\n',
+						Object.entries(sharedScope).map(([name, data]) =>
+							`${name}: ` +
+							Object.entries(data).map(
+								([version, data]) => `(${version}: ${data.from})`
+                            )
+						).join('\n ')
+					);
+
+					Promise.resolve(
+						container.init(sharedScope)
+					)
+					.then(
+						() => container.get(path)
+					)
+					.then(
+						(moduleFactory) => {
+							finalizeContainerRequest(
+								containerRequest, moduleFactory());
+						}
+					);
 				}
 				else {
 					const message =
-						'Container ' +
-						containerId +
-						' was fetched but ' +
-						'did not define itself in Liferay.Webpack.Container';
+						`Container ${containerId} was fetched but its script ` +
+						`failed to register with Liferay`;
 
 					console.warn(message);
 
-					reject(new Error(message));
+					finalizeContainerRequest(
+						containerRequest, new Error(message));
 				}
-			};
+			}
+		);
 
-			document.body.appendChild(script);
+
+		subscribeContainerRequest(
+			containerRequests[containerId], resolve, reject);
+	}
+
+	/**
+	 * Finalize an ongoing container request: set its result (error or module)
+	 * and notify all pending subscribers.
+	 */
+	function finalizeContainerRequest(containerRequest, result) {
+		const {script, subscribers} = containerRequest;
+
+		containerRequest.fetched = true;
+		containerRequest.script = undefined;
+		containerRequest.subscribers = undefined;
+
+		if(result instanceof Error) {
+			explain('Rejecting container', script.src, 'for', subscribers.length, 'subscribers');
+			containerRequest.error = result;
+			subscribers.forEach(({reject}) => reject(result));
+		} else {
+			explain('Resolving container', script.src, 'for', subscribers.length, 'subscribers');
+			containerRequest.module = result;
+			subscribers.forEach(({resolve}) => resolve(result));
 		}
-	});
+	};
+
+	/**
+	 * Get a defined container object
+	 */
+	function getContainer(containerId) {
+		return window[CONTAINERS_SYMBOL][containerId];
+	}
+
+	/**
+	 * Get a module (like require for webpack federation)
+	 */
+	function getModule(moduleName) {
+		return new Promise((resolve, reject) => {
+			const {containerId, path} = splitModuleName(moduleName);
+
+			const containerRequest = containerRequests[containerId];
+
+			if (!containerRequest) {
+				fetch(containerId, path, resolve, reject);
+			}
+			else {
+				const {error, module} = containerRequest;
```

Comment:
This is wrong because it assumes that one container can only have one module, which is wrong. It is working now because we only use one module per container now.